opm/autodiff/WellStateFullyImplicitBlackoil.hpp

```diff
@@ -282,6 +293,196 @@ namespace Opm
         }
 
 
+        /// init the MS well related.
+        template <typename PrevWellState>
+        void initWellStateMSWell(const Wells* wells, const std::vector<const Well*>& wells_ecl,
+                                 const int time_step, const PhaseUsage& pu, const PrevWellState& prev_well_state)
+        {
+            // still using the order in wells
+            const int nw = wells->number_of_wells;
+            if (nw == 0) {
+                return;
+            }
+
+            top_segment_loc_.clear();
+            top_segment_loc_.reserve(nw);
+            segpress_.clear();
+            segpress_.reserve(nw);
+            segrates_.clear();
+            segrates_.reserve(nw * numPhases());
+
+            nseg_ = 0.;
+            // in the init function, the well rates and perforation rates have been initialized or copied from prevState
+            // what we do here, is to set the segment rates and perforation rates
+            for (int w = 0; w < nw; ++w) {
+                const int nw_wells_ecl = wells_ecl.size();
+                int index_well_ecl = 0;
+                const std::string well_name(wells->name[w]);
+                for (; index_well_ecl < nw_wells_ecl; ++index_well_ecl) {
+                    if (well_name == wells_ecl[index_well_ecl]->name()) {
+                        break;
+                    }
+                }
+
+                // It should be able to find in wells_ecl.
+                if (index_well_ecl == nw_wells_ecl) {
+                    OPM_THROW(std::logic_error, "Could not find well " << well_name << " in wells_ecl ");
+                }
+
+                const Well* well_ecl = wells_ecl[index_well_ecl];
+                top_segment_loc_.push_back(nseg_);
+                if ( !well_ecl->isMultiSegment(time_step) ) { // not multi-segment well
+                    nseg_ += 1;
+                    segpress_.push_back(bhp()[w]);
+                    const int np = numPhases();
+                    for (int p = 0; p < np; ++p) {
+                        segrates_.push_back(wellRates()[np * w + p]);
+                    }
+                } else { // it is a multi-segment well
+                    const SegmentSet& segment_set = well_ecl->getSegmentSet(time_step);
+                    // assuming the oder of the perforations in well_ecl is the same with Wells
```

Comment:
oder -> order

Reply:
Also: if the assumption actually matters, it could be partially tested further down:
```
assert (nperf == wells.well_connpos[w+1] - wells.well_connpos[w]
```

Reply:
Ignore the above comment, I see there is an assert() further down. Sorry for the noise!